### DISCUSSION TITLE
Schema validation for mixed array type

### DISCUSSION THREAD

**Discussion by supercoffee** [TypedProject/ts-express-decorators#2781] [Q&A]
I'm trying to figure out the correct way to validate an array that can contain mixed object types while also validating that each of the objects matches a valid schema. I've been able to pull off mixed type validation with single properties, but this array is throwing me  off. 

Example data.  The last element in the array should not pass validation, but the first two should match the `ObjectWithId` and `UserWithEmail` schema respectively. 

```json
[
  {
    "_id": "asdf12345"
  },
  {
    "email": "user@example.com"
  },
  {
    "invalid": "invalid"
  }
]
```

Example schema

```typescript

@Description("Look up an existing resource by internal ID.")
export class ObjectWithId {
    @Property()
    _id: string;
}

@Description("Look up an existing user by email address.")
export class UserWithEmail {
    @Property()
    @Email()
    email: string;
}

class RequestBody {
    @Property()
    @Optional()
    @Schema(array().any(ObjectWithId, UserWithEmail))
    notify?: Array<ObjectWithId | UserWithEmail>;
}
```

I've tried all kinds of variations of `items`/`anyOf`/`oneOf` but nothing generates a schema other than 

```json
{
  "properties": {
    "notify" : {
      "type": "array"
    }
  }
}

**Comment by Romakita** [comment]
```
@Any(Model1,Model2)
prop: (Model1 | Model2)[]
```

See you

**Reply by supercoffee** [reply]
Hmm, that definitely get me closer, but not quite.  Now I see the correct schema definition, however there's no error for the "invalid" item.  The invalid item is neither removed from the array, nor does it trigger a validation error. 



Note: I'm using `@Optional` as I want to maintain compatibility for clients who aren't aware of this property.  Removing the `@Property` and `@Optional` decorators has no effect on the validation outcome. 

Using the following decorators
```
    @Property()
    @Optional()
    @Any(ObjectWithId, UserWithEmail)
    notify?: (ObjectWithId | UserWithEmail)[];
```

Generates the following schema.   To my untrained eyes, the schema looks correct so I'm not sure why the "invalid" item still passes validation.

```json
{
  "type": "array",
  "items": {
    "anyOf": [
      {
        "$ref": "#/definitions/ObjectWithId"
      },
      {
        "$ref": "#/definitions/UserWithEmail"
      }
    ]
  }
}
```

**Reply by Romakita** [reply]
Property and optionable isn't necessary when you use at least a schema decorator.

You can try OneOf instead Any (AnyOf).

Also you can test json schema behavior here:

https://www.jsonschemavalidator.net

This tool is helpful to understand  json schema behavior ;)

See you
